pkg/loadbalancer/reflectors/file.go

```diff
@@ -0,0 +1,230 @@
+)
+
+type fileReflectorConfig struct {
+	LBStateFile         string        `mapstructure:"lb-state-file"`
+	LBStateFileInterval time.Duration `mapstructure:"lb-state-file-interval"`
+}
+
+func (_ fileReflectorConfig) Flags(flags *pflag.FlagSet) {
+	flags.String("lb-state-file", "", "Synchronize load-balancing state from the specified file")
+	flags.Duration("lb-state-file-interval", time.Second, "Amount of time to wait before processing a changed load-balancing state file")
+	flags.MarkHidden("lb-state-file-interval")
+}
+
+func registerFileReflector(log *slog.Logger, jg job.Group, lbcfg loadbalancer.Config, extcfg loadbalancer.ExternalConfig, cfg fileReflectorConfig, w *writer.Writer) error {
+	if cfg.LBStateFile == "" || !lbcfg.EnableExperimentalLB {
+		return nil
+	}
+
+	file := cfg.LBStateFile
+	isYAML := false
+	switch {
+	case strings.HasSuffix(file, ".yaml"), strings.HasSuffix(file, ".yml"):
+		isYAML = true
+	case strings.HasSuffix(file, ".json"):
+		isYAML = false
+	default:
+		return fmt.Errorf("could not deduce file format from %q, expected .json or .yaml suffix", file)
+	}
+	ls := fileReflector{
+		log:       log,
+		cfg:       cfg,
+		extConfig: extcfg,
+		file:      file,
+		isYAML:    isYAML,
+		w:         w,
+	}
+	jg.Add(job.OneShot(
+		"sync",
+		ls.syncLoop,
+		job.WithRetry(-1, &job.ExponentialBackoff{
+			Min: time.Minute,
+			Max: 5 * time.Minute,
+		}),
+	))
+	return nil
+}
+
+// fileReflector syncs Source=LocalAPI state from a file containing a
+// [structs.LoadBalancerState].
+type fileReflector struct {
+	log       *slog.Logger
+	cfg       fileReflectorConfig
+	extConfig loadbalancer.ExternalConfig
+	file      string
+	isYAML    bool
+	w         *writer.Writer
+}
+
+func (s *fileReflector) syncLoop(ctx context.Context, health cell.Health) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	// Watch the parent directory for changes and react to events on the specified
+	// file.
+	watchDir := "."
+	if path.IsAbs(s.file) {
+		watchDir = path.Dir(s.file)
+	}
+	if err := watcher.Add(watchDir); err != nil {
+		return fmt.Errorf("failed to watch %q: %w", s.file, err)
+	}
+
+	limiter := rate.NewLimiter(s.cfg.LBStateFileInterval, 1)
+	defer limiter.Stop()
+
+	for {
+		if state, err := s.load(); err != nil {
+			s.log.Error("Failed to synchronize",
+				logfields.File, s.file,
+				logfields.Error, err)
+		} else {
+			txn := s.w.WriteTxn()
+			if numServices, numFrontends, numBackends, err := s.synchronize(txn, state); err == nil {
+				txn.Commit()
+				s.log.Info("Synchronized from local state",
+					logfields.File, s.file,
+					logfieldServiceCount, numServices,
+					logfieldFrontendCount, numFrontends,
+					logfieldBackendCount, numBackends)
+			} else {
+				s.log.Error("Failed to synchronize",
+					logfields.File, s.file,
+					logfields.Error, err)
+				txn.Abort()
+			}
+		}
+
+		for {
+			select {
+			case ev := <-watcher.Events:
```

Comment:
This isn't quite right. We should wait for the first event for the file and then wait `LBStateFileInterval` for further events. What happens now is that we might get multiple events about it (especially if editing the file with an editor) and we'd just reload multiple times.